### PR TITLE
Improved handling of env reload request - handles older version of worker sdk

### PR DESCRIPTION
**Pull request by kshyju**
Updating native placeholder to handle specialization request for function app payload which has older version worker package.

We will be updating the worker.sdk package to include a new property called ~"isSpecializable"~ `canUsePlaceholder` the "worker.config.json". The native placeholder will look for this property and if not present, it will the environment reload response with "failure" as status and a special exception. On the host we will check the exception associated with the failed env reload response and if it is this special exception. we will execute the fallback code - which is to shut down this channel and start a new channel with the customer payload dll.

For reference - the host changes are going to look like this: https://github.com/Azure/azure-functions-host/pull/9602 - This is the host version I used to test my changes locally.
 
### Pull request checklist

* [x] My changes **do not** require documentation changes
  * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
  * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
  * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

### CODE REVIEW

File: host/src/FunctionsNetHost/Grpc/WorkerConfig/WorkerDescription.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+namespace FunctionsNetHost.Grpc
+{
+    public sealed class WorkerDescription
+    {
+        public string? DefaultWorkerPath { set; get; }
+
+        public bool IsSpecializable { set; get; }
```
Comment (kshyju) [inline host/src/FunctionsNetHost/Grpc/WorkerConfig/WorkerDescription.cs:10]: Will be updating the worker.sdk package to include this new property in the worker.config.

File: host/src/FunctionsNetHost/Grpc/WorkerConfig/WorkerConfigUtils.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System.Text.Json;
+
+namespace FunctionsNetHost.Grpc
+{
+    internal static class WorkerConfigUtils
+    {
+        /// <summary>
+        /// Build and returns an instance of <see cref="WorkerConfig"/> from the worker.config.json file if present in the application directory.
```
Comment (satvu) [inline host/src/FunctionsNetHost/Grpc/WorkerConfig/WorkerConfigUtils.cs:11]:
```suggestion
        /// Builds and returns an instance of <see cref="WorkerConfig"/> from the worker.config.json file if present in the application directory.
```

**Review by satvu — APPROVED** [review]

File: host/src/FunctionsNetHost/Grpc/WorkerConfig/WorkerConfigUtils.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+                    Logger.Log($"worker.config.json not found at {workerConfigPath}. This may indicate missing app payload.");
+                    return null;
+                }
+
+                var jsonString = File.ReadAllText(workerConfigPath);
+                return JsonSerializer.Deserialize(jsonString, WorkerConfigSerializerContext.Default.WorkerConfig);
+            }
+            catch (Exception ex)
+            {
+                Logger.Log($"Error in WorkerConfigUtils.GetWorkerConfig.{ex}");
```
Comment (jviau) [inline host/src/FunctionsNetHost/Grpc/WorkerConfig/WorkerConfigUtils.cs:31]: Does this need a verbosity at all?

```suggestion
                Logger.Log($"Error in WorkerConfigUtils.GetWorkerConfig. {ex}");
```

File: host/src/FunctionsNetHost/Grpc/IncomingGrpcMessageHandler.cs [hunk 4]
```diff
@@ -50,16 +51,26 @@ private async Task Process(StreamingMessage msg)
...
                     var envReloadRequest = msg.FunctionEnvironmentReloadRequest;
-                    var applicationExePath = PathUtils.GetApplicationExePath(envReloadRequest.FunctionAppDirectory);
-                    Logger.LogTrace($"application path {applicationExePath}");
 
-                    if (string.IsNullOrWhiteSpace(applicationExePath))
+                    var workerConfig = WorkerConfigUtils.GetWorkerConfig(envReloadRequest.FunctionAppDirectory);
+
+                    if (workerConfig?.Description is null)
+                    {
+                        responseMessage.FunctionEnvironmentReloadResponse = BuildFailedEnvironmentReloadResponse(new FunctionAppPayloadNotFoundException());
```
Comment (jviau) [inline host/src/FunctionsNetHost/Grpc/IncomingGrpcMessageHandler.cs:59]: Same here with exception stack trace not being set. Is that an issue?

**Review by jviau — APPROVED** [review]

File: host/src/FunctionsNetHost/Grpc/WorkerConfig/WorkerConfigUtils.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace FunctionsNetHost.Grpc
+{
+    internal static class WorkerConfigUtils
+    {
+        /// <summary>
+        /// Build and returns an instance of <see cref="WorkerConfig"/> from the worker.config.json file if present in the application directory.
+        /// </summary>
+        /// <param name="applicationDirectory">The directory where function app deployed payload is present.</param>
+        internal static WorkerConfig? GetWorkerConfig(string applicationDirectory)
```
[the remark this answers is not in the document]
Comment (jviau) [reply host/src/FunctionsNetHost/Grpc/WorkerConfig/WorkerConfigUtils.cs:14]: Another thought: does the explicit check need to happen? You could just catch `FileNotFoundException` up the stack and treat that as the indicator the app payload may be missing.

File: host/src/FunctionsNetHost/Grpc/WorkerConfig/WorkerConfigUtils.cs (hunk 2 above)
Comment (kshyju) [reply host/src/FunctionsNetHost/Grpc/WorkerConfig/WorkerConfigUtils.cs]: Fixed comment as suggested. Thank you!

File: host/src/FunctionsNetHost/Grpc/Exceptions/EnvironmentReloadUnsupportedException.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+namespace Microsoft.Azure.Functions.Worker
+{
+    /// <summary>
+    /// The exception that is thrown when the current function app payload does not support environment reload.
+    /// </summary>
+    public sealed class EnvironmentReloadUnsupportedException : Exception
```
[the remark this answers is not in the document]
Comment (kshyju) [reply host/src/FunctionsNetHost/Grpc/Exceptions/EnvironmentReloadUnsupportedException.cs]: Updated to inherit from `NotSupportedException`. We wanted specific exception (for this specific scenario), hence not used the NotSupported one.

File: host/src/FunctionsNetHost/Grpc/IncomingGrpcMessageHandler.cs [hunk 7]
```diff
@@ -50,16 +51,26 @@ private async Task Process(StreamingMessage msg)
...
+                        break;
+                    }
+
+                    // function app payload which uses an older version of Microsoft.Azure.Functions.Worker package does not support specialization.
+                    if (!workerConfig.Description.IsSpecializable)
                     {
-                        var ex = new InvalidOperationException($"Unable to find a valid function app payload at '{envReloadRequest.FunctionAppDirectory}'");
-                        responseMessage.FunctionEnvironmentReloadResponse = BuildFailedEnvironmentReloadResponse(ex);
+                        Logger.LogTrace("App payload uses an older version of worker package which does not support specialization.");
+                        responseMessage.FunctionEnvironmentReloadResponse = BuildFailedEnvironmentReloadResponse(new EnvironmentReloadUnsupportedException());
```
[the remark this answers is not in the document]
Comment (kshyju) [reply host/src/FunctionsNetHost/Grpc/IncomingGrpcMessageHandler.cs:67]: Yes. For now, We use this to signal the caller (host) that the operation failed due to this. Not filling all those properties is on purpose. We don't use any of the properties today. Can revisit this later.

File: host/src/FunctionsNetHost/Grpc/WorkerConfig/WorkerConfigUtils.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kshyju) [reply host/src/FunctionsNetHost/Grpc/WorkerConfig/WorkerConfigUtils.cs]: Yes, Good idea. In the latest iteration I pushed to the async version and switched to FileNotFoundException as suggested.

File: host/src/FunctionsNetHost/Grpc/WorkerConfig/WorkerConfigUtils.cs (hunk 3 above)
Comment (kshyju) [reply host/src/FunctionsNetHost/Grpc/WorkerConfig/WorkerConfigUtils.cs:34]: We don't have different overloads with verbosity now. the `Logger.Log` is a central place to write to std out as of today.

File: host/src/FunctionsNetHost/Grpc/Exceptions/EnvironmentReloadUnsupportedException.cs [hunk 8]
```diff
@@ -0,0 +1,15 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+namespace Microsoft.Azure.Functions.Worker
+{
+    /// <summary>
+    /// The exception that is thrown when the current function app payload does not support environment reload.
+    /// </summary>
+    public sealed class EnvironmentReloadUnsupportedException : NotSupportedException
```
Comment (brettsam) [inline host/src/FunctionsNetHost/Grpc/Exceptions/EnvironmentReloadUnsupportedException.cs:9]: Rename `EnvironmentReloadNotSupportedException` to match better with `NotSupportedException`

File: host/src/FunctionsNetHost/Grpc/WorkerConfig/WorkerDescription.cs (hunk 1 above)
Comment (brettsam) [inline host/src/FunctionsNetHost/Grpc/WorkerConfig/WorkerDescription.cs:10]: Rename `CanSpecialize`? Or `CanUsePlaceholder`?

**Review by brettsam — APPROVED** [review]

File: host/src/FunctionsNetHost/Grpc/WorkerConfig/WorkerDescription.cs (hunk 1 above)
Comment (kshyju) [reply host/src/FunctionsNetHost/Grpc/WorkerConfig/WorkerDescription.cs]: Switched to "CanUsePlaceholder"

File: host/src/FunctionsNetHost/Grpc/Exceptions/EnvironmentReloadUnsupportedException.cs (hunk 8 above)
Comment (kshyju) [reply host/src/FunctionsNetHost/Grpc/Exceptions/EnvironmentReloadUnsupportedException.cs]: Renamed as suggested.

**Pull request merged**